### PR TITLE
09_Count views

### PR DESCRIPTION
**Pull request by mmarchois**
PR 9

See https://github.com/poledev/mathieu/issues/16

### CODE REVIEW

File: app/config/config.yml [hunk 1]
```diff
@@ -90,3 +90,5 @@ stof_doctrine_extensions:
 
 knp_paginator:
     page_range: 5
+    template:
+        pagination: KnpPaginatorBundle:Pagination:sliding.html.twig
```
Comment (florianbelhomme) [inline app/config/config.yml]: C'était déjà mentionné dans une autre PR :/

Comment (mmarchois) [inline app/config/config.yml]: Elle n'est pas encore mergé

File: src/SensioLabs/JobBoardBundle/Manager/JobManager.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        $this->entityManager = $entityManagerInterface;
+    }
+
+    /**
+     * @param Job $job
+     */
+    public function increaseDetailViews(Job $job)
+    {
+        $job->setCountDetail($job->getCountDetail() + 1);
+        $this->entityManager->flush($job);
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Manager/JobManager.php]: [My comment on a previous commit](https://github.com/poledev/mathieu/commit/07b6fea8797e8564579a4088fbbf3eba857c1192#commitcomment-11628311) was hidden: Be careful about race conditions. Your implementation does not prevent it.

File: src/SensioLabs/JobBoardBundle/Controller/BaseController.php [hunk 3]
```diff
@@ -46,9 +50,11 @@ public function indexAction(Request $request, $page)
             'paginator' => $paginator,
             'countries' => $jobRepository->getJobsCountries(),
             'contracts' => $jobRepository->getJobsContractType(),
-            'countryFilter' => $request->get('country'),
-            'contractTypeFilter' => $request->get('contract-type'),
-            'filterValues' => $request->query->all(),
+            'filters' => [
+                'country' => $request->get('country'),
+                'contractType' => $request->get('contract-type'),
+                'allRequest' => $request->query->all()
```
Comment (salahm) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: `allRequest` what it means ?

File: src/SensioLabs/JobBoardBundle/Entity/Job.php [hunk 4]
```diff
@@ -117,11 +117,11 @@ class Job
     protected $valid;
 
     /**
-     * @ORM\Column(type = "integer", name = "count_show")
+     * @ORM\Column(type = "integer", name = "count_detail")
      *
      * @var integer
      */
-    protected $countShow;
+    protected $countDetail;
```
Comment (salahm) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: `countDetailView` ?

File: src/SensioLabs/JobBoardBundle/Controller/BaseController.php (hunk 3 above)
Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: It retrieve all request query

File: src/SensioLabs/JobBoardBundle/Tests/Entity/JobTest.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * {@inheritDoc}
+ */
+class JobTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function testGetTotalCount()
```
Comment (salahm) [inline src/SensioLabs/JobBoardBundle/Tests/Entity/JobTest.php]: the test should be a functionnal one !

File: src/SensioLabs/JobBoardBundle/Manager/JobManager.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+{
+    /**
+     * @var EntityManagerInterface
+     */
+    protected $entityManager;
+
+    /**
+     * @param EntityManagerInterface $entityManagerInterface
+     */
+    public function __construct(EntityManagerInterface $entityManagerInterface)
```
Comment (romqin) [inline src/SensioLabs/JobBoardBundle/Manager/JobManager.php]: Renommer `$entityManagerInterface` par `$entityManager` ?

File: src/SensioLabs/JobBoardBundle/Manager/JobManager.php (hunk 2 above)
Comment (skigun) [inline src/SensioLabs/JobBoardBundle/Manager/JobManager.php]: You can also use a nice setter for that like a `$job->IncrementCountDetail()` or whatever

File: src/SensioLabs/JobBoardBundle/Controller/BaseController.php (hunk 3 above)
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: `queryParameters` ?

**Comment by ogizanagi** [comment]
Trouver une moyen de régler le problème potentiel de race conditions et ça semble bon pour moi :+1:

File: src/SensioLabs/JobBoardBundle/Controller/BaseController.php [hunk 7]
```diff
@@ -33,8 +35,10 @@ public function indexAction(Request $request, $page)
     {
         $jobRepository = $this->getDoctrine()->getRepository('SensioLabsJobBoardBundle:Job');
 
-        $findJobs = $jobRepository->findJobs($request->query->all());
-        $paginator = $this->get('knp_paginator')->paginate($findJobs, $page, self::POST_PER_PAGE);
+        $query = $jobRepository->findJobs($request->query->all());
+        $paginator = $this->get('knp_paginator')->paginate($query, $page, self::POST_PER_PAGE);
+
+        $jobRepository->increaseCountViews('countListViews', $paginator->getItems());
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: You do not retrieve the updated entities

File: src/SensioLabs/JobBoardBundle/Tests/Entity/JobTest.php (hunk 5 above)
Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Tests/Entity/JobTest.php]: I have added a new functionnal test

File: src/SensioLabs/JobBoardBundle/Repository/JobRepository.php [hunk 8]
```diff
@@ -95,4 +94,42 @@ public function findMyJobs(User $user)
...
+     * @param string $field
+     * @param array  $jobs
+     */
+    public function increaseCountViews($field, array $jobs)
+    {
+        if (empty($jobs)) {
+            return;
+        }
+
+        if (!in_array($field, ['countDetailViews', 'countListViews', 'countApiViews'])) {
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Repository/JobRepository.php]: Shouldn't this throw an Exception in order to warn the user something anormal happened ?

**Pull request merged**